Add Dolphin resource pack support

AnErrupTion opened the pull request:
Hello there! :wave:

This PR adds support for Dolphin's resource pack feature, which allows for easy installation of game textures in Dolphin's GUI. A resource pack consists of a manifest file, a logo, and a `textures` folder which contains all the textures under the game ID's folder. What I've done is make this repository's folder structure pack-compliant, so that anybody can simply download the repository as a ZIP file, put it in Dolphin's `ResourcePacks` folder, then install it.

With that said, while this is unrelated to the actual PR, I haven't been able to install the resource pack with the PNG files. The DDS one works fine, but with the PNG one, Dolphin says it loads the textures, but the textures in-game look identical to the original game (also, there are nearly twice as many textures as in the DDS version, why is that?).

If the PNG version isn't designed to be used with Dolphin, perhaps it's a good idea to put the DDS textures under the `textures` folder, and the PNG ones in another folder in the repository.

Finally, for the logo, I simply chose the already existing `mkwii-reloaded.jpg` file. It's *definitely* not adapted for use as a logo, but I'm terrible at designing logos, so I figured this would be a good-enough placeholder for now.

AnErrupTion committed:
Add Dolphin resource pack support

Signed-off-by: AnErrupTion <anerruption@disroot.org>

GhostlyDark commented:
I have a couple of issues with this. It doesn't seem to work at all regardless of format, as Dolphin is unable to copy the textures from the ResourcePack into the `Textures/Load` directory, so the install step just won't work. That happens with both PNG and DDS files. Does it not work with portable installs?

Dolphin has a feature to allow the texture folder to be named whatever, with the supported game IDs being text files inside the `Load` folder of the texture pack. That also allows for multiple game IDs to be supported. It's unclear to me if that is something that works using ResourcePacks or not (I assume they do, since ResourcePacks are supposed to just be unzipped and deleted from the `Load/Textures` directory when installing or uninstalling via the GUI). If so, then renaming the folder to `RMC` wouldn't be required, even though the specs say otherwise (but they are 6 years old...so...). That thought would be more important for other packs though.

DDS files are converted from the source PNG files. Uploading DDS files to the GitHub repo is something to be avoided due to their file size and offering a ResourcePack in PNG format doesn't make much sense due to their performance impact. I also think that downloading the source from GitHub as a zip file won't work as a ResourcePack out of the box either, since it's wrapped in an extra directory, so Dolphin won't recognize the `manifest.json` that way anyhow.

All that being said, I am open to throw in a manifest and change the folder structure to accomodate for this, but the feature seems to be buggy from what I've seen.

AnErrupTion commented:
> I have a couple of issues with this. It doesn't seem to work at all regardless of format, as Dolphin is unable to copy the textures from the ResourcePack into the `Textures/Load` directory, so the install step just won't work. That happens with both PNG and DDS files. Does it not work with portable installs?

I don't see why it would work less on portable installs than on regular installs, unless Dolphin cannot locate the `Load` directory (but then, if it can't, one has bigger problems to deal with). I'm not sure where to get a portable version of Dolphin to try and reproduce this issue though, is it just a matter of copying the emulator's files to another directory?

> Dolphin has a feature to allow the texture folder to be named whatever, with the supported game IDs being text files inside the `Load` folder of the texture pack. That also allows for multiple game IDs to be supported. It's unclear to me if that is something that works using ResourcePacks or not (I assume they do, since ResourcePacks are supposed to just be unzipped and deleted from the `Load/Textures` directory when installing or uninstalling via the GUI). If so, then renaming the folder to `RMC` wouldn't be required, even though the specs say otherwise (but they are 6 years old...so...). That thought would be more important for other packs though.

Ah, so that's what this file is for. Initially installing the resource pack without renaming the folder didn't work, and I didn't think much about it and just went ahead and renamed it. But after taking a closer look, I came to realise that the `Load/RMC.txt` wasn't being extracted because it was empty. Adding a few characters inside the file seemed to fix the issue. Probably a bug that's worth reporting upstream. Should I rename the folder back then?

> I also think that downloading the source from GitHub as a zip file won't work as a ResourcePack out of the box either, since it's wrapped in an extra directory, so Dolphin won't recognize the `manifest.json` that way anyhow.

Indeed, I completely forgot about that! :sweat_smile:

mrpenguinb commented:
To make any 5.0+ version of Dolphin portable, just create a blank "portable.txt" file in the same folder as the .exe/application.
